### PR TITLE
Soft bundle basic observability

### PR DESCRIPTION
**Pull request by shio-coder**
## Description 

The PR adds several metrics and logging to Soft Bundle.

A node operator will have the ability to understand what bundles are submitted, as well as the content.

This can be particularly useful when tracing down a transaction.

## Test plan 

How did you test the new or updated feature?

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

**Review by mystenmark — COMMENTED** [review]
@williampsmith can you take a look as well since this touches the client IP extraction?

**Comment by williampsmith** [comment]
tl;dr - LGTM on the client IP extraction side. 

One thing to note is that the `client_id_source` field of the validator server is currently populated based on a node config setting for a specific feature that need not be enabled. So if we want to reuse it for logging in other scenarios such as in this PR, it would make sense to decouple that config from the aforementioned feature. I can do that in a separate PR though as it seems out of scope for this. Should be fine to reuse in the meantime.

File: crates/sui-core/src/authority_server.rs [hunk 1]
```diff
@@ -909,6 +942,98 @@ impl ValidatorService {
         Ok((tonic::Response::new(response), Weight::one()))
     }
 
+    fn get_client_ip_addr<T>(&self, request: &tonic::Request<T>) -> Option<IpAddr> {
+        if self.client_id_source.is_none() {
+            return None;
+        }
```
Comment (williampsmith) [inline crates/sui-core/src/authority_server.rs:948]: You may consider defaulting to the behavior of the `ClientIdSource::SocketAddr` variant case shown below (line 952-971) here, as perhaps something is better than nothing (which will be the case 95% of the time as most nodes will not have the necessary node config for `client_id_source` to be set. The reason the code you're referencing does an early return if `client_id_source` is not set is because the code needs to know that the IP it is collecting is indeed the IP of the originating client (or fullnode, in the validator case) and not some intermediate proxy. That is, nothing is better than something incorrect in that case. Feel free to use your best judgement here given this context

Comment (shio-coder) [reply crates/sui-core/src/authority_server.rs:948]: Thanks for the input!

I guess one possible way is to add an argument to `get_client_ip_addr` to control this behavior (whether defaulting to the said variant), keep reusing this method, so that this will work for now.  Does that make sense?

Comment (shio-coder) [reply crates/sui-core/src/authority_server.rs]: That makes perfect sense.  I have updated the PR according to this.

**Comment by shio-coder** [comment]
Hi @mystenmark ,

Guess we are almost done with the client IP extraction part.
Let me know if there are more comments.

Thank you

File: crates/sui-core/src/authority_server.rs [hunk 2]
```diff
@@ -453,22 +465,30 @@ impl ValidatorService {
 
         let shared_object_tx = certificates[0].contains_shared_object();
```
Comment (mwtian) [inline crates/sui-core/src/authority_server.rs:466]: I think we should check if any certificate contains shared object.

File: crates/sui-core/src/authority_server.rs [hunk 3]
```diff
@@ -453,22 +465,30 @@ impl ValidatorService {
...
-                .start_timer()
+            if wait_for_effects {
+                &self
+                    .metrics
+                    .handle_soft_bundle_certificates_consensus_latency
+            } else {
+                &self
+                    .metrics
+                    .handle_soft_bundle_certificates_non_consensus_latency
+            }
```
Comment (mwtian) [inline crates/sui-core/src/authority_server.rs:487]: This branch should check if any transaction in the bundle contains a shared object. And  I don't think we need to add a metric corresponding to `submit_certificate_consensus_latency` (`!wait_for_effects`) for soft bundle for now.

Comment (shio-coder) [reply crates/sui-core/src/authority_server.rs]: I have added a comment here saying that the check has already been ensured by `soft_bundle_validity_check`, so in practice all certificates will contain shared objects when entering this branch.  Since that is the case, also removed `handle_soft_bundle_certificates_non_consensus_latency`.

File: crates/sui-core/src/authority_server.rs [hunk 4]
```diff
@@ -827,6 +838,7 @@ impl ValidatorService {
         request: tonic::Request<HandleSoftBundleCertificatesRequestV3>,
     ) -> WrappedServiceResponse<HandleSoftBundleCertificatesResponseV3> {
         let epoch_store = self.state.load_epoch_store_one_call_per_task();
+        let client_addr = self.get_client_ip_addr(&request, &ClientIdSource::SocketAddr);
```
Comment (williampsmith) [inline crates/sui-core/src/authority_server.rs:841]: Almost. Per our discussion this should be the folllowing:

```suggestion
        let client_addr = if self.client_id_source.is_none() {
            self.get_client_ip_addr(&request, &ClientIdSource::SocketAddr)
        } else {
            self.get_client_ip_addr(&request, client_id_source.unwrap())
        };
```

File: crates/sui-core/src/authority_server.rs [hunk 5]
```diff
@@ -840,6 +856,18 @@ impl ValidatorService {
         self.soft_bundle_validity_check(&certificates, &epoch_store)
             .await?;
 
+        info!(
```
Comment (mwtian) [inline crates/sui-core/src/authority_server.rs:859]: Usually Sui logs at debug level per request / transaction. But this can be fine because of the low initial TPS.

**Comment by williampsmith** [comment]
Thanks @shio-coder. Looks good on my side. Will leave final stamp to @mwtian

**Review by mwtian — APPROVED** [review]
LGTM

**Comment by shio-coder** [comment]
Thank you both @williampsmith and @mwtian !   Really appreciate the effort on reviewing this.
